src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,877 @@
+          setVoxel(Mn::Vector3i(i, ind, k), "negYShadow", true);
+        } else if (getVoxel<bool>(Mn::Vector3i(i, ind, k), "Boundary")) {
+          hit = true;
+          setVoxel(Mn::Vector3i(i, ind, k), "negYShadow", true);
+        }
+      }
+      // fill posX
+      hit = false;
+      ind = -1;
+      while (++ind < m_voxelGridDimensions[1]) {
+        if (hit) {
+          setVoxel(Mn::Vector3i(i, ind, k), "posYShadow", true);
+        } else if (getVoxel<bool>(Mn::Vector3i(i, ind, k), "Boundary")) {
+          hit = true;
+          setVoxel(Mn::Vector3i(i, ind, k), "posYShadow", true);
+        }
+      }
+    }
+  }
+  // Z axis ray casts
+  for (int i = 0; i < m_voxelGridDimensions[0]; i++) {
+    for (int j = 0; j < m_voxelGridDimensions[1]; j++) {
+      // fill negX
+      hit = false;
+      ind = m_voxelGridDimensions[2] - 1;
+      while (ind--) {
+        if (hit) {
+          setVoxel(Mn::Vector3i(i, j, ind), "negZShadow", true);
+        } else if (getVoxel<bool>(Mn::Vector3i(i, j, ind), "Boundary")) {
+          hit = true;
+          setVoxel(Mn::Vector3i(i, j, ind), "negZShadow", true);
+        }
+      }
+      // fill posX
+      hit = false;
+      ind = -1;
+      while (++ind < m_voxelGridDimensions[2]) {
+        if (hit) {
+          setVoxel(Mn::Vector3i(i, j, ind), "posZShadow", true);
+        } else if (getVoxel<bool>(Mn::Vector3i(i, j, ind), "Boundary")) {
+          hit = true;
+          setVoxel(Mn::Vector3i(i, j, ind), "posZShadow", true);
+        }
+      }
+    }
+  }
+
+  // create int grid
+  addGrid<int>(gridName);
+  bool nX = false, pX = false, nY = false, pY = false, nZ = false, pZ = false;
+  int hash = 0;
+  // fill in int grid with voting approach
+  for (int i = 0; i < m_voxelGridDimensions[0]; i++) {
+    for (int j = 0; j < m_voxelGridDimensions[1]; j++) {
+      for (int k = 0; k < m_voxelGridDimensions[2]; k++) {
+        Mn::Vector3i index = Mn::Vector3i(i, j, k);
+        if (getVoxel<bool>(index, "Boundary")) {
+          setVoxel(index, gridName, 0);
+          continue;
+        }
+        nX = !getVoxel<bool>(index, "negXShadow");
+        pX = !getVoxel<bool>(index, "posXShadow");
+        nY = !getVoxel<bool>(index, "negYShadow");
+        pY = !getVoxel<bool>(index, "posYShadow");
+        nZ = !getVoxel<bool>(index, "negZShadow");
+        pZ = !getVoxel<bool>(index, "posZShadow");
+        // || ((nX || pX) && (nY || pY) && (nZ || pZ))
+        if (((nX && pX) || (nY && pY) || (nZ && pZ)) ||
+            ((nX || pX) && (nY || pY) && (nZ || pZ))) {
+          // Exterior (+inf)
+          setVoxel(index, gridName, INT_MAX);
+        } else {
+          // Interior (-inf)
+          setVoxel(index, gridName, INT_MIN);
+        }
+      }
+    }
+  }
+  // remove the transient utility grids
+  removeGrid("negXShadow");
+  removeGrid("posXShadow");
+  removeGrid("negYShadow");
+  removeGrid("posYShadow");
+  removeGrid("negZShadow");
+  removeGrid("posZShadow");
```

Comment:
Probably could be worth to have a way to create a grid but don't add it to the map (which then means it doesn't need to have any string name, can get removed/destructed implicitly at the end of scope etc...)